server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -494,9 +515,13 @@ public SnapshotInfo finalizeSnapshot(final SnapshotId snapshotId,
             startTime, failure, System.currentTimeMillis(), totalShards, shardFailures,
             includeGlobalState, userMetadata);
         try {
+            final BlobContainer indicesBlobContainer = blobStore().blobContainer(basePath().add("indices"));
+            final Map<String, BlobContainer> foundIndices = indicesBlobContainer.children();
+            final RepositoryData updatedRepositoryData = getRepositoryData().addSnapshot(snapshotId, blobStoreSnapshot.state(), indices);
+            final var survivingIndices = updatedRepositoryData.getIndices();
             snapshotFormat.write(blobStoreSnapshot, blobContainer(), snapshotId.getUUID());
-            final RepositoryData repositoryData = getRepositoryData();
-            writeIndexGen(repositoryData.addSnapshot(snapshotId, blobStoreSnapshot.state(), indices), repositoryStateId);
+            writeIndexGen(updatedRepositoryData, repositoryStateId);
+            cleanupStaleIndices(foundIndices, survivingIndices);
```

Comment:
as discussed, let's not do the clean-up as part of the snapshotting, but only during deletion.
Instead let's think about how we can add a way for the user to force a clean even if there are no snapshots to be deleted.

Reply:
Ok, will revert that. Unfortunately this will also force doing something about the new repo consistency tests (those were reliant on the fact that the last snapshot taken would clean up the repo). I'll think of something there :)